src/lib/System/IO/Streams/SequenceId.hs

```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE TupleSections #-}
+
+module System.IO.Streams.SequenceId
+       ( sequenceIdInputStream
+       , sequenceIdOutputStream
+       ) where
+
+import           Control.Applicative ((<$>))
+import           Data.IORef          (newIORef, readIORef, writeIORef, atomicModifyIORef')
+import           Data.SequenceId     (SequenceIdError, checkSeqId,
+                                      incrementSeqId)
+import           System.IO.Streams   (InputStream, OutputStream)
+import qualified System.IO.Streams   as Streams
+
+
+------------------------------------------------------------------------------
+-- | Wrap an 'System.IO.Streams.InputStream' and check for dropped or duplicated sequence IDs.
+--
+-- Example:
+--
+-- @
+-- ghci> is <- 'System.IO.Streams.fromList' [1..10::Int]
+-- ghci> (is', resetSeqId) <- 'sequenceIdInputStream' 0 id (fail . show) is
+-- ghci> 'System.IO.Streams.read' is'
+-- Just 1
+-- ghci> 'System.IO.Streams.read' is'
+-- Just 2
+-- ghci> 'System.IO.Streams.read' is'
+-- Just 3
+-- ghci> resetSeqId 0
+-- 3
+-- ghci> 'System.IO.Streams.read' is'
+-- *** Exception: user error ('Data.SequenceId.SequenceIdError' {errType = 'Data.SequenceId.SequenceIdDropped', lastSeqId = 0, currSeqId = 4})
+-- @
+sequenceIdInputStream :: Integral s
+                      => s                             -- ^ Initial sequence ID
+                      -> (a -> s)                      -- ^ Function applied to each element of the stream to get the sequence ID
+                      -> (SequenceIdError s -> IO ())  -- ^ Error handler
+                      -> InputStream a                 -- ^ 'System.IO.Streams.InputStream' to check the sequence of
+                      -> IO (InputStream a, s -> IO s) -- ^ Pass-through of the given stream, and 'IO' action that returns
+                                                       -- the current sequence id and then resets it to the initial seed
+sequenceIdInputStream initSeqId getSeqId seqIdFaultHandler =
+    inputFoldM f initSeqId
+  where
+    f lastSeqId x = do
+        let currSeqId = getSeqId x
+        maybe (return ()) seqIdFaultHandler $ checkSeqId lastSeqId currSeqId
```

Comment:
`maybe (return ())` is `traverse_`, right?

Reply:
yep, fixed!